### PR TITLE
Feature/report service client

### PR DESCRIPTION
**Pull request by esotrope**

### CODE REVIEW

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/security/GroupGrant.java [hunk 1]
```diff
@@ -7,14 +7,23 @@
...
 
-    private final long id;
-    private final int subjectId;
+    private long id;
+    private int subjectId;
+
+    /**
+     * No-arg contructor for JSON serialization and deserialization
+     */
+    private GroupGrant(){}
```
Comment (esotrope) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/security/GroupGrant.java]: Needed for JSON deserialization

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/ReportWebServiceClient.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+/**
+ * This class acts as an interface to the report web service API
+ */
+@Service
+@SuppressWarnings({"unchecked"})
+class ReportWebServiceClient {
```
Comment (bgsavage) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/ReportWebServiceClient.java]: No interface?  Or still a todo?

Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/ReportWebServiceClient.java]: Haven't been making interfaces for the WS clients but yeah we probably should.

File: pdf-report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/ReportController.java [hunk 3]
```diff
@@ -67,33 +64,29 @@ public Report generateReport(@RequestBody final BatchExamReportRequestHolder req
...
-            response.setHeader(CONTENT_DISPOSITION, "attachment; filename=" + report.getLabel() + ".pdf");
-            IOUtils.copy(pdfBytes, responseOutput);
-            response.flushBuffer();
-        }
+
+        final InputStreamResource resource = new InputStreamResource(service.getContent(reportId));
+        final HttpHeaders headers = new HttpHeaders();
+        headers.add(CONTENT_TYPE, "application/pdf; charset=UTF-8");
+        headers.add(CONTENT_DISPOSITION, "attachment; filename=" + report.getLabel() + ".pdf");
+        return new ResponseEntity<>(resource, headers, HttpStatus.OK);
```
Comment (esotrope) [inline pdf-report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/ReportController.java]: @wtritch springy version of the same thing.

Comment (wtritch) [inline pdf-report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/web/ReportController.java]: I don't think that'll work.  Check out AbstractResource#contentLength()

File: pdf-report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/web/ReportControllerIT.java [hunk 4]
```diff
@@ -134,15 +134,14 @@ public void itShouldRetrieveAListOfReportsByUserAndId() throws Exception {
...
 
-        when(service.getContent(123))
+        when(service.getContent(report.getId()))
                 .thenAnswer((id) -> new ByteArrayInputStream("PDF CONTENTS".getBytes()));
 
         mvc.perform(get("/api/reports/123/content"))
-
                 .andExpect(status().isOk())
-                .andExpect(header().string(CONTENT_TYPE, APPLICATION_PDF_VALUE))
+                .andExpect(header().string(CONTENT_TYPE, "application/pdf; charset=UTF-8"))
```
Comment (wtritch) [inline pdf-report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/web/ReportControllerIT.java]: Interesting, needed for UTF-8 compliance?

File: webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/DefaultReportWebServiceClient.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        final ResponseEntity<List<Report>> response = template.exchange(
+                "{url}/api/reports?user={userId}", HttpMethod.GET, null,
+                new ParameterizedTypeReference<List<Report>>() {}, url, userId);
+        return getBodyOrThrowExceptionIfNotSuccessful(response);
+    }
+
+    @Override
+    public void streamReport(final String userId, final long reportId, final OutputStream outputStream) {
+        template.execute("{url}/api/reports/{reportId}/content?user={userId}",
+                HttpMethod.GET, request -> {}, new StreamExtractor(outputStream), url, reportId, userId);
```
Comment (esotrope) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/DefaultReportWebServiceClient.java]: @wtritch added streaming - The StreamExtractor passed here and defined below is what allows us to write the rest template input stream directly to an output stream.

Comment (wtritch) [inline webapp/src/main/java/org/opentestsystem/rdw/reporting/report/client/DefaultReportWebServiceClient.java]: Nice, and Spring doesn't read the stream into a buffer in order to calculate the Content-Length header?

**Pull request merged**